book_source/source/2darrays.md

```diff
@@ -0,0 +1,237 @@
+```
+[[0 , 1 , 2 , 3 , 4],
+[5 , 6 , 7 , 8 , 9 ],
+[10, 11, 12, 13, 14],
+[15, 16, 17, 18, 19],
+[20, 21, 22, 23, 24]]
+```
+Notice there is an extra pair of opening and closing brackets, which represents the array that stores all of the other arrays. The syntax is commonly represented as array[row][col] and is represented visually as so above for clarity. Additionally, the call to an array using the syntax array[1] would still function, only instead of providing a primative type, it would instead provide the array stored at that index of the array of arrays.
+
+````{tab-set-code}
+
+```{code-block} java
+int[][] array = new int[5][5];
+array[1][2] = 1;
+array[1][3] = 14;
+array[3][4] = array[1][2];
+array[4][4] = 5;
+
+for (int i = 0; i < array.length; i++) {
+    System.out.println(Arrays.toString(array[i]));
+}
+```
+
+```{code-block} python
+array = [[0]*5] *5  # Short-hand to make a 2D length 5 list
+array[1][2] = 1;
+array[1][3] = 14;
+array[3][4] = array[1][2];
+array[4][4] = 5;
+
+print(array)
+```
+
+```{code-block} javascript
+let array = [[0,0,0,0,0], [0,0,0,0,0], [0,0,0,0,0], [0,0,0,0,0], [0,0,0,0,0]];
+
+array[1][2] = 1;
+array[1][3] = 14;
+array[3][4] = array[1][2];
+array[4][4] = 5;
+
+console.log(array);
+```
+
+```{code-block} c
+int array[5][5] = {{0}};
+array[1][2] = 1;
+array[1][3] = 14;
+array[3][4] = array[1][2];
+array[4][4] = 5;
+
+printf("[");
+for (int i = 0; i < 5; i++) {
+for (int j = 0; j < 5; j++) {
+printf("%d, ", array[i][j]);
+}
+printf("\n");
+}
+```
+
+```{code-block} c++
+int array[5][5] = {
+    {0, 0, 0, 0, 0},
+    {0, 0, 0, 0, 0},
+    {0, 0, 0, 0, 0},
+    {0, 0, 0, 0, 0},
+    {0, 0, 0, 0, 0},
+};
+
+array[1][2] = 1;
+array[1][3] = 14;
+array[3][4] = array[1][2];
+array[4][4] = 5;
+for (int i = 0; i < 5; ++i) {
+for (int j = 0; j < 5; ++j) {
+std::cout << array[i][j] << " ";
+}
+std::cout << std::endl; 
+}
+```
+````
+
+Similarily, if we want to change the value at (3,4) of the 2D array to 1, where 3 is the "x" coordinate or outside array and 4 is the "y" or inner array we would do something like array[3][4] = 1. Note that the 2D array can only be of one type, just like how a 1D array is only one type.
+
+## Looping over 2D Arrays
+
+We can also use a for loop to view and/or modify each index, although since this is a nested data structure (meaning there is a data structure within another data structure), we need to use a nested for loop in return. The starting value for each loop will typically start at 0, going to either array.length or array[0].length. Note the second for loop, we use array[0].length since we only want to loop through to the end of the inner array length, represented by array[0].
```

Comment:
Shortening this a bit - I think the walkthrough of the initialization value and tests isn't strictly necessary, since it's evident in the code right below. 
```suggestion
To traverse over a 2D array, it's very common to use a _nested for loop structure_ where the outer for loops iterates over each _row_ of the 2D array, and the inner for loop iterates over each _element_ within each row. See some examples below. 
```